pkgs/desktops/gnome-3/extensions/update-extensions.py

```diff
@@ -0,0 +1,136 @@
+    remote = urllib.request.urlopen(url)
+    hasher = hashlib.sha256()
+
+    while True:
+        data = remote.read(4096)
+        if not data:
+            break
+        hasher.update(data)
+    return hasher.hexdigest()
+
+
+# Fetching extensions
+page = 0
+extensions = []
+while True:
+    page += 1
+    print("Scraping page " + str(page))
+    try:
+        with urllib.request.urlopen(
+            f"https://extensions.gnome.org/extension-query/?n_per_page=25&page={page}"
+        ) as url:
+            data = json.loads(url.read().decode())["extensions"]
+            responseLength = len(data)
+            print(f"Found {responseLength} extensions")
+
+            for extension in data:
+                # Yeah, there are some extensions without any releases
+                if not extension["shell_version_map"]:
+                    continue
+
+                print("Processing " + extension["uuid"])
+
+                # Throw away that 'pk' key. What is it for?
+                extension["shell_version_map"] = {
+                    k: v["version"] for k, v in extension["shell_version_map"].items()
+                }
+                # Transform shell_version_map to be more useful for us. Also throw away unwanted versions
+                extension["shell_version_map"] = generate_extension_versions(
+                    extension["shell_version_map"], extension["uuid"]
+                )
+                if not extension["shell_version_map"]:
+                    # No compatible versions found
+                    continue
+
+                # Parse something like "/extension/1475/battery-time/" and output "battery-time-1475"
+                def pname_from_url(url: str) -> str:
+                    url = url.split("/")  # type: ignore
+                    return url[3] + "-" + url[2]
+
+                extension["pname"] = pname_from_url(extension["link"])
+                extension["link"] = "https://extensions.gnome.org" + extension["link"]
+
+                # Replace \u0123 in strings. This is required for nix < 2.3.10
+                # extension["name"] = extension["name"].encode('unicode-escape').decode()
+                # extension["description"] = extension["description"].encode('unicode-escape').decode()
+
+                # Remove unused keys
+                del extension["screenshot"]
+                del extension["icon"]
+                del extension["creator"]
+                del extension["creator_url"]
+                del extension["pk"]
+
+                extensions += [extension]
+
+            if responseLength < 25:
+                break
+    except urllib.error.HTTPError as e:
+        print("Got error. We're done. (This should be a 404)\n" + str(e))
+        break
+
+print(f"Writing results ({len(extensions)} extensions in total)")
+with open("extensions.json", "w") as out:
+    # Manually pretty-print the outer level, but then do one compact line per extension
```

Comment:
Having been part of this discussion I know why you are doing this. It is worth explaining that in one more sentence.

You could also add a check that ensures that it is still valid JSON after writing the file. Just to be sure you aren't creating an automation that automated generating garbage.

Reply:
Wouldn't this be immediately checked by the CI building the extensions anyways?

Reply:
Maybe, maybe not. Would be better to just *never* generate invalid JSON documents. Someone that doesn't know all the details might be adding another field and then things fail with bad Nix error messages instead of a proper error message that we can generate here.